core/common/src/main/java/alluxio/security/authentication/ChannelAuthenticator.java

```diff
@@ -0,0 +1,148 @@
+      AuthType authType) {
+    mUseSubject = false;
+    mChannelId = UUID.randomUUID();
+    mUserName = userName;
+    mPassword = password;
+    mImpersonationUser = impersonationUser;
+    mAuthType = authType;
+  }
+
+  /**
+   * Authenticates given {@link NettyChannelBuilder} instance. It attaches required interceptors to
+   * the channel based on authentication type.
+   *
+   * @param managedChannel the managed channel for whch authentication is taking place
+   * @return channel that is augmented for authentication
+   * @throws AuthenticationException
+   */
+  public Channel authenticate(ManagedChannel managedChannel) throws AuthenticationException {
+    if (mAuthType == AuthType.NOSASL) {
+      return managedChannel;
+    }
+
+    // Create a channel for talking with target host's authentication service.
+    try {
+      // Create SaslClient for authentication based on provided credentials.
+      SaslClient saslClient;
+      if (mUseSubject) {
+        saslClient =
+            SaslParticipiantProvider.Factory.create(mAuthType).createSaslClient(mParentSubject);
+      } else {
+        saslClient = SaslParticipiantProvider.Factory.create(mAuthType).createSaslClient(mUserName,
+            mPassword, mImpersonationUser);
+      }
+
+      // Create authentication scheme specific handshake handler.
+      SaslHandshakeClientHandler handshakeClient =
+          SaslHandshakeClientHandler.Factory.create(mAuthType, saslClient);
+      // Create driver for driving sasl traffic from client side.
+      SaslStreamClientDriver clientDriver = new SaslStreamClientDriver(handshakeClient);
+      // Start authentication call with the service and update the client driver.
+      StreamObserver<SaslMessage> requestObserver =
+          SaslAuthenticationServiceGrpc.newStub(managedChannel).authenticate(clientDriver);
+      clientDriver.setServerObserver(requestObserver);
+      // Start authentication traffic with the target.
+      clientDriver.start(mChannelId.toString());
+      // Authentication succeeded!
+      // Attach scheme specific interceptors to the channel.
+
+      Channel authenticatedChannel =
+          ClientInterceptors.intercept(managedChannel, getInterceptors(saslClient));
+      return authenticatedChannel;
+
+    } catch (UnauthenticatedException e) {
+      throw new AuthenticationException(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * @param saslClient the Sasl client object that have been used for authentication
+   * @return the list of interceptors that are required for configured authentication
+   */
+  private List<ClientInterceptor> getInterceptors(SaslClient saslClient) {
+    if (!SecurityUtils.isSecurityEnabled()) {
+      return Collections.emptyList();
+    }
+    List<ClientInterceptor> interceptorsList = new ArrayList<>();
+    AuthType authType =
+        Configuration.getEnum(PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.class);
```

Comment:
Should we be checking `mAuthType` here?

Reply:
What do you mean?

Reply:
We get the `AuthType` from the configuration, but we already have a field `mAuthType` that the class was instantiated with?

Reply:
Makes sense.